### PR TITLE
CBL-4238 : Pass cert chain to the root_cert_locator()

### PR DESCRIPTION
**Pull request by pasin**
* Ported the fix from release/3.1 branch (6fde3088edd41fa549dced003ba01a41fe0697cc) which is for CBL-4237.

Background :

In LiteCore PublicKey+Apple.mm’s findSigningRootCert() for TLS CA cert validation, a SecTrust is created with only a leaf cert without its intermediate certs. As a result, when calling SecTrustEvaluateWithError() to evaluate the SecTrust (a technique to get all certs including the root CA cert used for verifying a CA signed cert), there is an error logged as follows:

```
[TLS] error: SecTrustEvaluateWithError failed: Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed:  / The attempt to create a certificate chain failed."
```

Fix:

* This commit is constructing and passing the cert chain (in PEM format) to the root_cert_locator(). The passed cert chain, will eventually be passed to the findSigningRootCert() that Apple platform can use when constructing an array of SecCertRef for creating a SecTrust. Note that there will be a corresponding fix in LiteCore regarding this as well.

* Use vector<char> as a dynamic buffer with an initial buffer size of 10000 which should be big enough for most cases. When the buffer size is not big enough, resize the buffer with the min buffer size set in olen out param by the mbedtls_pem_write_buffer function and retry again.

### CODE REVIEW

File: src/mbedtls_context.cpp [hunk 1]
```diff
@@ -585,7 +587,38 @@ namespace sockpp {
...
+            int ret = 0;
+            size_t olen = 10000; // initial buffer size
+            std::vector<unsigned char> buf;
+            for (int i = 0; i < 2; i++) {
+                buf.resize(olen);
+                ret = mbedtls_pem_write_buffer("-----BEGIN CERTIFICATE-----\n",
+                                               "-----END CERTIFICATE-----\n",
+                                               crt->raw.p, crt->raw.len,
+                                               buf.data(), buf.size(), &olen);
+                if (ret != MBEDTLS_ERR_BASE64_BUFFER_TOO_SMALL) {
```
Comment (bmeike) [inline src/mbedtls_context.cpp:603]: Is this correct?  It looks like it only runs the loop twice if the buffer isn't big enough??

**Review by bmeike — APPROVED** [review]
Looks ok.  I just have a question about the handling of a MBEDTLS_ERR_BASE64_BUFFER_TOO_SMALL err.

File: src/mbedtls_context.cpp (hunk 1 above)
Comment (bmeike) [reply src/mbedtls_context.cpp:603]: NVM.... I get it.  `mbedtls_pem_write_buffer` will set olen to the size that it needs, if it does not have enough room?

**Review by jianminzhao — APPROVED** [review]

File: src/mbedtls_context.cpp (hunk 1 above)
Comment (pasin) [reply src/mbedtls_context.cpp:603]: Yes, the olen will get updated with the min buffer size to use.

**Pull request merged**